File: skills/faion/knowledge/dev/go-error-handling/templates/error-middleware.go
// purpose: HTTP middleware translating returned-error AppError into JSON via errors.As.
// consumes: handlers of type `func(w,r) error`; AppError from pkg/apperror.
// produces: JSON response with Content-Type application/json (or problem+json if integrated with error-handling methodology).
// depends-on: pkg/apperror, encoding/json, net/http.
// token-budget-impact: ~70 lines; loaded once at boot.
// Package httpx provides HTTP handler wrappers including error translation middleware.
// Handler is a func(ResponseWriter, *Request) error variant of http.HandlerFunc.
// Wrap converts it to a standard HandlerFunc and translates *AppError to JSON responses.
package httpx

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"

	"yourmod/pkg/apperror"
)

type errBody struct {
	Code    string `json:"code"`
	Message string `json:"message"`
	TraceID string `json:"trace_id,omitempty"`
}

// Handler is like http.HandlerFunc but returns an error.
type Handler func(http.ResponseWriter, *http.Request) error

// Wrap converts a Handler to http.HandlerFunc with error translation.
func Wrap(h Handler) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		err := h(w, r)
		if err == nil {
			return
		}
		var appErr *apperror.AppError
		if !errors.As(err, &appErr) {
			appErr = apperror.Wrap(err, "internal error")
		}
		if appErr.HTTPStatus >= 500 {
			slog.ErrorContext(r.Context(), "http error",
				"code", appErr.Code, "status", appErr.HTTPStatus, "err", err)
		} else {
			slog.InfoContext(r.Context(), "http client error",
				"code", appErr.Code, "status", appErr.HTTPStatus)
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(appErr.HTTPStatus)
		if encErr := json.NewEncoder(w).Encode(errBody{
			Code:    appErr.Code,
			Message: appErr.Message,
			TraceID: TraceIDFromCtx(r.Context()),
		}); encErr != nil {
			slog.ErrorContext(r.Context(), "failed to encode error response", "err", encErr)
		}
	}
}

// TraceIDFromCtx extracts the trace ID from context — implement per project.
func TraceIDFromCtx(r interface{ Value(any) any }) string {
	if id, ok := r.Value("trace_id").(string); ok {
		return id
	}
	return ""
}

File: skills/faion/knowledge/dev/nodejs-express/templates/error-middleware.ts
// purpose: centralised error mapping middleware for Express
// consumes: known error kinds enumerated by the service
// produces: deterministic JSON error responses; no stack leakage
// depends-on: express; pino for logging
// token-budget-impact: ~160 tokens when loaded as context
import type { ErrorRequestHandler } from 'express'

const STATUS_MAP: Record<string, number> = {
  ValidationError: 400,
  AuthError: 401,
  ForbiddenError: 403,
  NotFoundError: 404,
  ConflictError: 409,
  RateLimitError: 429,
}

export const errorMiddleware: ErrorRequestHandler = (err, req, res, _next) => {
  const status = STATUS_MAP[err?.name] ?? err?.status ?? 500
  const code = err?.code ?? err?.name ?? 'internal'
  req.log?.error({ err, status, code }, 'request failed')
  res.status(status).json({ error: code, message: status >= 500 ? 'internal' : err?.message })
}
